pytorch_lightning/trainer/connectors/training_trick_connector.py

```diff
@@ -28,10 +28,21 @@ def on_trainer_init(
         gradient_clip_algorithm: str,
         track_grad_norm: Union[int, float, str],
         accumulate_grad_batches: Union[int, Dict[int, int]],
-        terminate_on_nan: bool,
+        terminate_on_nan: Optional[bool],
```

Comment:
quick question. To me, terminate_on_nan meaning is slightly different from detect_anomaly. You could detect an anomaly where a nan loss was created and decide to keep training while just ignoring it.
Do you think this replacement is required ? Or could we find a better term ?

Reply:
We actually had a feature to enable training with NaN Loss, which could be quite useful for un-stable losses like CTCLoss: https://github.com/PyTorchLightning/pytorch-lightning/pull/5359

Reply:
The infinite loss check is called here:
```def backward_fn(loss: Tensor):
    self.backward(loss, optimizer, opt_idx)

    # when in dev debugging track the losses
    # TODO: remove dev debugger tracking loss history
    self.trainer.dev_debugger.track_train_loss_history(batch_idx, loss)

    # check if loss or model weights are nan
    if self.trainer.terminate_on_nan:
        check_finite_loss(self.trainer.lightning_module, loss)

    return loss
```
We could add:
```
if self.trainer.terminate_on_nan or self.trainer.detect_anomaly:
    check_finite_loss(self.trainer.lightning_module, loss)
```
However, to allow users to continue training with NaN loss, we would have to add a separate trainer argument or not enable `check_finite_loss` when using `detect_anomaly`.

Reply:
Why would an user whose model/loss produces NaN losses by design set `detect_anomaly`?

I don't think this is a big deal, we can add support if it gets requested

Reply:
haven't followed all discussions here but in case it wasn't clear, terminate on nan has two purposes:

- when loss is not finite, stop training before running backward -> weights don't get contaminated and user knows the bug is in the loss function
- check if weights are nan or inf after backward -> if they are, means backward produced them so user can start debugging there

if `detect_anomaly` can cover this behavior we don't need terminate_on_nan anymore and can just rename/remove it. 
I agree with @carmocca detect_anomaly/terminate_on_nan is for debugging and should not be turned on when NaN anomalies are expected by design.